iothub_client/src/uamqp_messaging.c

```diff
@@ -430,79 +443,219 @@ int message_create_uamqp_encoding_from_iothub_message(IOTHUB_MESSAGE_HANDLE mess
+                        LogError("Failed to get value of uAMQP message 'message-id' property (string)");
+                        string_value = NULL;
+                        result = __FAILURE__;
+                    }
+                }
+                else if (value_type == AMQP_TYPE_ULONG)
+                {
+                    uint64_t ulong_value;
+
+                    // Codes_SRS_UAMQP_MESSAGING_09_014: [The message-id value shall be retrieved from the AMQP_VALUE as char sequence] 
+                    if (amqpvalue_get_ulong(uamqp_message_property, &ulong_value) != 0)
+                    {
+                        // Codes_SRS_UAMQP_MESSAGING_09_015: [If message-id fails to be obtained, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.] 
+                        LogError("Failed to get value of uAMQP message 'message-id' property (ulong)");
+                        string_value = NULL;
+                        result = __FAILURE__;
+                    }
+                    else if (sprintf(string_buffer, "%" PRIu64, ulong_value) < 0)
+                    {
+                        LogError("Failed converting 'message-id' (ulong) to string");
+                        string_value = NULL;
+                        result = __FAILURE__;
+                    }
+                    else
+                    {
+                        string_value = string_buffer;
+                    }
+                }
+                else if (value_type == AMQP_TYPE_UUID)
+                {
+                    uuid uuid_value;
+
+                    // Codes_SRS_UAMQP_MESSAGING_09_014: [The message-id value shall be retrieved from the AMQP_VALUE as char sequence] 
+                    if (amqpvalue_get_uuid(uamqp_message_property, &uuid_value) != 0)
+                    {
+                        // Codes_SRS_UAMQP_MESSAGING_09_015: [If message-id fails to be obtained, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.] 
+                        LogError("Failed to get value of uAMQP message 'message-id' property (UUID)");
+                        string_value = NULL;
+                        result = __FAILURE__;
+                    }
+                    else if ((string_value = UUID_to_string((UUID*)uuid_value)) == NULL)
+                    {
+                        // Codes_SRS_UAMQP_MESSAGING_09_015: [If message-id fails to be obtained, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.] 
+                        LogError("Failed to get the string representation of 'message-id' UUID");
+                        string_value = NULL;
+                        result = __FAILURE__;
+                    }
+                    else
+                    {
+                        (void)memcpy(string_buffer, string_value, sizeof(char) * strlen(string_value));
```

Comment:
>memcpy [](start = 30, length = 6)

Also - why is this a memcpy?  Shouldn't it just be a straight strcpy at this point?

Reply:
Strcpy goes pretty well when I don’t want to deal or don’t know the size of the string, and I’m counting on the '\0' as terminator.

Since here I know the size of the string, memcpy is preferable as behind the scenes the function doesn’t need to be searching for the ‘\0’ throughout the char sequence.

Plus, there are lots of talks about how strcpy can be “evil”. So I advocate to keep memcpy.